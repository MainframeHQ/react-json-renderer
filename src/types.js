// @flow

import type { Component, Element } from 'react'

type Props = { [key: string]: mixed }

type StatelessComponent = (props: Props) => Element<*>

export type ComponentType = string | ReactComponent

export type ReactComponent = StatelessComponent | Component<*, *, *>

export type ComponentChild = void | string | ComponentConfig

export type ComponentChildren = ComponentChild | Array<ComponentChild>

export type ComponentProps = Props & {
  children?: ComponentChildren,
}

export type ComponentConfig = {
  type: ComponentType,
  props: ComponentProps,
}

export type ConvertedChild = void | string | ConvertedComponent

export type ConvertedChildren = ConvertedChild | Array<ConvertedChild>

export type ConvertedProps = Props & {
  children?: ConvertedChildren,
}

export type ConvertedComponent = {
  type: string,
  props: ConvertedProps,
}
