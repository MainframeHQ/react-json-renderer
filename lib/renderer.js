'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Renderer = exports.renderFromObject = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var Fallback = function Fallback() {
  return null;
};

var renderFromObject = function renderFromObject(tree, params) {
  var components = params && params.components || {};
  var fallback = params && params.fallback || Fallback;

  var createChild = function createChild(c) {
    if (c == null) {
      return null;
    }
    if (typeof c === 'string') {
      return c;
    }
    if (c.type) {
      return createComponent(c);
    }
    return null;
  };

  var createComponent = function createComponent(converted) {
    var component = components[converted.type] || fallback;

    var _converted$props = converted.props,
        children = _converted$props.children,
        props = _objectWithoutProperties(_converted$props, ['children']);

    if (children) {
      props.children = Array.isArray(children) // eslint-disable-line react/prop-types
      ? children.map(createChild) : createChild(children);
    }
    return (0, _react.createElement)(component, props);
  };

  return createComponent(tree);
};

exports.renderFromObject = renderFromObject;

var Renderer = exports.Renderer = function (_Component) {
  _inherits(Renderer, _Component);

  function Renderer() {
    _classCallCheck(this, Renderer);

    return _possibleConstructorReturn(this, (Renderer.__proto__ || Object.getPrototypeOf(Renderer)).apply(this, arguments));
  }

  _createClass(Renderer, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          json = _props.json,
          tree = _props.tree,
          params = _objectWithoutProperties(_props, ['json', 'tree']);

      var obj = void 0;
      if (tree) obj = tree;else if (json) obj = JSON.parse(json);

      return obj ? renderFromObject(obj, params) : null;
    }
  }]);

  return Renderer;
}(_react.Component);