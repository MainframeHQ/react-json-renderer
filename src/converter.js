// @flow

import React from 'react'

import type {
  ComponentChild,
  ComponentChildren,
  ComponentConfig,
  ComponentProps,
  ComponentType,
  ConvertedChild,
  ConvertedChildren,
  ConvertedComponent,
  ReactComponent,
} from './types'

type Replacer = (props: ComponentProps) => ComponentConfig
type ConvertParams = {
  replacers?: { [key: string]: Replacer },
}

export const convertToObject = (
  tree: ComponentConfig,
  params?: ConvertParams = {},
): ConvertedComponent => {
  const replacers = params.replacers || {}

  const convertChild = (child: ComponentChild): ConvertedChild => {
    if (child == null) {
      return
    }
    if (typeof child === 'string') {
      return child
    }
    if (child.type) {
      return convertComponent(child)
    }
  }

  const convertChildren = (children: ComponentChildren): ConvertedChildren => {
    return Array.isArray(children)
      ? children.map(convertChild)
      : convertChild(children)
  }

  const convertComponent = (tree: ComponentConfig, key?: string) => {
    if (typeof tree.key === 'string') {
      key = tree.key
    }

    let name, type
    if (typeof tree.type === 'string') {
      name = tree.type
      type = 'string'
    } else if (
      typeof tree.type.prototype === 'object' &&
      tree.type.prototype &&
      tree.type.prototype.isReactComponent
    ) {
      name = tree.type.displayName || tree.type.name
      type = 'class'
    } else if (typeof tree.type === 'function') {
      name = tree.type.displayName || tree.type.name || 'Unknown'
      type = 'function'
    } else {
      return {
        type: 'Unsupported',
        props: {
          children: [],
        },
      }
    }

    const replacer = replacers[name]
    if (replacer) {
      return convertComponent(replacer(tree.props), key)
    }

    let children
    if (type === 'string') {
      children = convertChildren(tree.props.children)
    } else if (type === 'function') {
      children = convertChildren(tree.type(tree.props))
    } else if (type === 'class') {
      const instance = new tree.type(tree.props)
      children = convertChildren(instance.render())
    }

    return {
      type: name,
      props: { ...tree.props, children, key },
    }
  }

  return convertComponent(tree)
}

type ConvertJSONParams = ConvertParams & {
  space?: number | string,
}

export const convertToJSON = (
  tree: ComponentConfig,
  params?: ConvertJSONParams = {},
): string => {
  const { space, ...convertParams } = params
  return JSON.stringify(convertToObject(tree, convertParams), null, space)
}
