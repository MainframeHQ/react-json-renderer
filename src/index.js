// @flow

export { convertToObject, renderToJSON } from './converter'
export { renderFromObject, Renderer } from './renderer'
