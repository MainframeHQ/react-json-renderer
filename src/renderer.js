// @flow

import React, { createElement, PureComponent, type ComponentType } from 'react'

import type { ConvertedElement } from './types'

type RenderParams = {
  components?: { [type: string]: ComponentType<*> },
  fallback?: ComponentType<*>,
}

const Fallback = () => null

export const renderFromObject = (
  tree: ConvertedElement,
  params?: RenderParams,
) => {
  const components = (params && params.components) || {}
  const fallback = (params && params.fallback) || Fallback

  const createChild = c => {
    if (c == null) {
      return null
    }
    if (Array.isArray(c)) {
      return c.map(createChild)
    }
    if (
      typeof c === 'boolean' ||
      typeof c === 'number' ||
      typeof c === 'string'
    ) {
      return c
    }
    if (c.type) {
      return createFromObject(c)
    }
    return null
  }

  const createFromObject = (converted: ConvertedElement) => {
    const component = components[converted.type] || fallback

    if (!converted.props) {
      return createElement(component)
    }

    const { children, ...props } = converted.props
    if (children != null) {
      props.children = createChild(children)
    }
    return createElement(component, props)
  }

  return createFromObject(tree)
}

type RendererProps = {
  json?: string,
  tree?: ConvertedElement,
} & RenderParams

export class Renderer extends PureComponent<RendererProps> {
  props: RendererProps

  render() {
    const { json, tree, ...params } = this.props
    let obj
    if (tree) obj = tree
    else if (json) obj = JSON.parse(json) // flowlint-line sketchy-null-string:off

    return obj ? renderFromObject(obj, params) : null
  }
}
