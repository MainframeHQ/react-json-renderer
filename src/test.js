import React from 'react'

import { renderToJSON, renderFromObject } from './index'

const Empty = () => null
const View = ({ children }) => children
const Hello = ({ name }) => `Hello ${name}`
const Placeholder = 'Placeholder'
class ClassComponent extends React.Component {
  render() {
    return this.props.children
  }
}

const tree = (
  <ClassComponent>
    <View style="primary">
      <Empty />
      <Hello name="Bob" />
      <View><Placeholder /></View>
    </View>
  </ClassComponent>
)

const JSONRes = renderToJSON(tree)

const objectRes = JSON.parse(JSONRes)

const components = {
  ClassComponent: ({ children }) => children,
  Empty: () => null,
  View: ({ children }) => children,
  Hello: ({ name }) => `Hello ${name}`,
  Placeholder: () => 'Placeholder text',
}

const res = renderFromObject(objectRes, { components })

console.log('res', res)
