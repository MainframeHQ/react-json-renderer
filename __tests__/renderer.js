/* global describe it */

import React, { Component, Fragment } from 'react'
import renderer from 'react-test-renderer'

import { renderFromObject, Renderer } from '../src/renderer'

const Fallback = () => <span>fallback!</span>
const Text = ({ value }) => <span>{value}</span>
const View = ({ children }) => <div>{children}</div>

const components = { Text, View }

const tree1 = {
  type: 'Test',
  props: {
    children: ['test'],
  },
}

const tree2 = {
  type: 'View',
  props: {
    children: [
      {
        type: 'Text',
        props: {
          key: 'text',
          value: 'hello',
          children: ['world'],
        },
      },
      {
        type: 'Image',
        props: {
          key: 'img',
          source: 'URI',
        },
      },
    ],
  },
}

describe('renderFromObject()', () => {
  it('returns the default fallback if no components and fallback are provided', () => {
    expect(renderFromObject(tree1)).toMatchSnapshot()
  })

  it('applies the provided components and fallback', () => {
    expect(
      renderFromObject(tree2, {
        components,
        fallback: Fallback,
      }),
    ).toMatchSnapshot()
  })

  it('returns null for unsupported children', () => {
    expect(
      renderFromObject(
        {
          type: 'Text',
          props: { children: ['test', undefined, 2, null, 'other', true] },
        },
        { components },
      ),
    ).toMatchSnapshot()
  })
})

describe('Renderer', () => {
  it('returns null if no components and fallback are provided', () => {
    const tree = renderer.create(<Renderer tree={tree1} />).toJSON()
    expect(tree).toBe(null)
  })

  it('returns the processed tree', () => {
    const tree = renderer
      .create(
        <Renderer tree={tree2} components={components} fallback={Fallback} />,
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('handles not having props provided', () => {
    const tree = renderer
      .create(
        <Renderer
          tree={{ type: 'View' }}
          components={components}
          fallback={Fallback}
        />,
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('supports nesting children arrays', () => {
    const tree = renderer
      .create(
        <Renderer
          tree={{
            type: 'View',
            props: {
              children: [
                [{ type: 'Text', props: { key: '1', value: 'Hello' } }],
                [
                  { type: 'Text', props: { key: '2', value: ' nested ' } },
                  [[{ type: 'Text', props: { key: '3', value: 'children' } }]],
                ],
              ],
            },
          }}
          components={components}
          fallback={Fallback}
        />,
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('supports Fragment', () => {
    const tree = renderer
      .create(
        <Renderer
          tree={{
            type: 'Fragment',
            props: {
              children: [
                { type: 'Text', props: { key: '1', value: 'Hello' } },
                { type: 'Text', props: { key: '2', value: 'world' } },
              ],
            },
          }}
          components={components}
          fallback={Fallback}
        />,
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('parses and renders the provided JSON string', () => {
    const json = JSON.stringify(tree1)
    const tree = renderer
      .create(<Renderer json={json} components={{ Test: View }} />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
