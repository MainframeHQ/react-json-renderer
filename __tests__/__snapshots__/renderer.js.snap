// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renderer handles not having props provided 1`] = `<div />`;

exports[`Renderer parses and renders the provided JSON string 1`] = `
<div>
  test
</div>
`;

exports[`Renderer returns the processed tree 1`] = `
<div>
  <span>
    hello
  </span>
  <span>
    fallback!
  </span>
</div>
`;

exports[`renderFromObject() applies the provided components and fallback 1`] = `
<View>
  <Text
    value="hello"
  >
    world
  </Text>
  <Fallback
    source="URI"
  />
</View>
`;

exports[`renderFromObject() returns null for unsupported children 1`] = `
<Text>
  test
  2
  other
  true
</Text>
`;

exports[`renderFromObject() returns the default fallback if no components and fallback are provided 1`] = `
<Fallback>
  test
</Fallback>
`;
